pkg/kotsadm/push_images.go

```diff
@@ -332,14 +332,214 @@ func TagAndPushAppImages(imagesDir string, options types.PushImagesOptions) ([]k
 	return images, nil
 }
 
+func TagAndPushAppImagesFromBundle(airgapBundle string, options types.PushImagesOptions) ([]kustomizetypes.Image, error) {
+	if options.LogForUI {
+		writeProgressLine(options.ProgressWriter, "Reading image information from bundle...")
+	}
+
+	imageFiles, err := getImageListFromBundle(airgapBundle, options.LogForUI)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get layer info from bundle")
+	}
+
+	fileReader, err := os.Open(airgapBundle)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to open file")
+	}
+	defer fileReader.Close()
+
+	gzipReader, err := gzip.NewReader(fileReader)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get new gzip reader")
+	}
+	defer gzipReader.Close()
+
+	reportWriter := options.ProgressWriter
+	if options.LogForUI {
+		wc := reportWriterWithProgress(imageFiles, options.ProgressWriter)
+		reportWriter = wc.(io.Writer)
+		defer wc.Write([]byte(fmt.Sprintf("+status.flush:\n")))
+		defer wc.Close()
+	}
+
+	images := []kustomizetypes.Image{}
+
+	tarReader := tar.NewReader(gzipReader)
+	for {
+		header, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to get read archive")
+		}
+
+		if header.Typeflag != tar.TypeReg {
+			continue
+		}
+
+		imageFile, ok := imageFiles[header.Name]
+		if !ok {
+			continue
+		}
+
+		err = func() error {
+			pathParts := strings.Split(imageFile.FilePath, string(os.PathSeparator))
+			if len(pathParts) < 3 {
+				return errors.Errorf("not enough path parts in %q", imageFile.FilePath)
+			}
+
+			rewrittenImage, err := image.ImageInfoFromFile(options.Registry, pathParts[2:])
+			if err != nil {
+				return errors.Wrap(err, "failed to decode image from path")
+			}
+
+			if options.LogForUI {
+				writeProgressLine(reportWriter, fmt.Sprintf("Extracting image %s:%s", rewrittenImage.NewName, rewrittenImage.NewTag))
+			}
+
+			tmpFile, err := ioutil.TempFile("", "kotsadm-app-image-")
+			if err != nil {
+				return errors.Wrap(err, "failed to create temp file")
+			}
+			defer tmpFile.Close()
+			defer os.Remove(tmpFile.Name())
+
+			_, err = io.Copy(tmpFile, tarReader)
+			if err != nil {
+				return errors.Wrapf(err, "failed to write file %q", header.Name)
+			}
+
+			tmpFile.Close()
```

Comment:
there's already a `defer tmpFile.Close()` on line 405, not sure i get why this is added here

Reply:
this is to flush all data before passing file name to containers package